Enhancement/8172 show view only users popup

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8172 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationIntroductoryOverlayNotification.js:
```diff
@@ -58,6 +61,20 @@ export default function AudienceSegmentationIntroductoryOverlayNotification() {
+	const scrollToWidgetAndDismissNotification = ( event ) => {
+		event.preventDefault();
+
+		const widgetClass =
+			'.googlesitekit-widget-area--mainDashboardTrafficPrimary';
```
I realised this is as per the IB, but in fact we should scroll to the audience segmentation widget area:

```suggestion
		const widgetAreaClass =
			'.googlesitekit-widget-area--mainDashboardTrafficAudienceSegmentation';
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationIntroductoryOverlayNotification.js:
```diff
@@ -83,7 +100,7 @@ export default function AudienceSegmentationIntroductoryOverlayNotification() {
 					tertiary
 					disabled={ isDismissing }
-					onClick={ dismissNotification }
+					onClick={ scrollToWidgetAndDismissNotification }
```
The click handler is applied to the wrong button, this is the "Got it" button, it should be applied to the "Show me" button.

techanvil commented on assets/js/components/OverlayNotification/OverlayNotificationsRenderer.js:
```diff
@@ -24,23 +24,30 @@ import { Fragment } from '@wordpress/element';
 
 export default function OverlayNotificationsRenderer() {
 	const audienceSegmentationEnabled = useFeature( 'audienceSegmentation' );
+	const audienceSegmentationSetupComplete = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getAudienceSegmentationSetupComplete()
+	);
```
This results in the banner being shown for all users, whereas we only want to show it for secondary users i.e. not for the admin who initially set the feature up. We missed a few details in the IB.

In fact we are going to need to make a change to the `getAudienceSegmentationSetupComplete()` setting so that instead of simply storing a boolean, it should store the ID of the user who set the feature up, so we can then check here if the ID is a match for the current user. It's a bit of a pain but we should rename `getAudienceSegmentationSetupComplete()` to `getAudienceSegmentationSetupCompletedBy()` at the same time, to make it semantically clear.

So, please can you apply these changes to the setting, but also rename it to `getAudienceSegmentationSetupCompletedBy()`. You'll need to update a few tests as a result, it's unfortunate but we do need to do it here.

```diff
diff --git a/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.js b/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.js
index f7f41a78b0..4d8a407aa8 100644
--- a/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.js
+++ b/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.js
@@ -193,7 +193,7 @@ function AudienceSegmentationSetupCTAWidget( { Widget, WidgetNull } ) {
 	}
 
 	if (
-		audienceSegmentationSetupComplete ||
+		audienceSegmentationSetupComplete !== null ||
 		configuredAudiences === undefined ||
 		configuredAudiences?.length ||
 		! analyticsIsDataAvailableOnLoad ||
diff --git a/assets/js/modules/analytics-4/datastore/audiences.js b/assets/js/modules/analytics-4/datastore/audiences.js
index a910222a32..2c40456f1a 100644
--- a/assets/js/modules/analytics-4/datastore/audiences.js
+++ b/assets/js/modules/analytics-4/datastore/audiences.js
@@ -522,8 +522,10 @@ const baseActions = {
 			return { error };
 		}
 
+		const userID = select( CORE_USER ).getID();
+
 		dispatch( MODULES_ANALYTICS_4 ).setAudienceSegmentationSetupComplete(
-			true
+			userID
 		);
 
 		const { saveSettingsError } = yield commonActions.await(
diff --git a/assets/js/modules/analytics-4/index.js b/assets/js/modules/analytics-4/index.js
index 42833473f1..a7b5098c1a 100644
--- a/assets/js/modules/analytics-4/index.js
+++ b/assets/js/modules/analytics-4/index.js
@@ -175,7 +175,7 @@ export const registerWidgets = ( widgets ) => {
 				return (
 					availableAudiences?.length &&
 					configuredAudiences === null &&
-					audienceSegmentationSetupComplete === true
+					audienceSegmentationSetupComplete !== null
 				);
 			},
 		},
diff --git a/includes/Modules/Analytics_4/Settings.php b/includes/Modules/Analytics_4/Settings.php
index 1531a67c77..3c74f3d8a4 100644
--- a/includes/Modules/Analytics_4/Settings.php
+++ b/includes/Modules/Analytics_4/Settings.php
@@ -107,7 +107,7 @@ class Settings extends Module_Settings implements Setting_With_Owned_Keys_Interf
 			'adsLinkedLastSyncedAt'             => 0,
 			'availableAudiences'                => null,
 			'availableAudiencesLastSyncedAt'    => 0,
-			'audienceSegmentationSetupComplete' => false,
+			'audienceSegmentationSetupComplete' => null,
 			'detectedEvents'                    => array(),
 		);
 	}
@@ -208,7 +208,9 @@ class Settings extends Module_Settings implements Setting_With_Owned_Keys_Interf
 				}
 
 				if ( isset( $option['audienceSegmentationSetupComplete'] ) ) {
-					$option['audienceSegmentationSetupComplete'] = (bool) $option['audienceSegmentationSetupComplete'];
+					if ( ! is_int( $option['audienceSegmentationSetupComplete'] ) ) {
+						$option['audienceSegmentationSetupComplete'] = null;
+					}
 				}
 			}
```

Once the above is done, you should update this condition as follows. However one more thing to consider is that it's starting to add quite a bit of logic specific to the `AudienceSegmentationIntroductoryOverlayNotification` component into `OverlayNotificationsRenderer`. So it might be preferable to move this condition into `AudienceSegmentationIntroductoryOverlayNotification` for cleaner encapsulation of its logic. That being the case you could remove the tests you've added for `OverlayNotificationsRenderer`.

```suggestion
	const shouldShowAudienceSegmentationIntroductoryOverlay = useSelect(
		( select ) =>
			select( CORE_MODULES ).isModuleActive( 'analytics-4' ) &&
			( ! isViewOnly ||
				select( CORE_USER ).canViewSharedModule( 'analytics-4' ) ) &&
			select(
				MODULES_ANALYTICS_4
			).getAudienceSegmentationSetupCompletedBy() !==
				select( CORE_USER ).getID()
	);
```

techanvil commented on assets/js/components/OverlayNotification/OverlayNotificationsRenderer.test.js:
Note that if we do move the condition for rendering `AudienceSegmentationIntroductoryOverlayNotification` into the component itself as per my comment on the issue, this entire file could be removed and won't need an update. Please review that comment before addressing those below.

techanvil reviewed (changes requested):
Hi @ankitrox, thanks for the work so far on this. However we have unfortunately missed a couple of things in the IB and this needs some additional work. Please see my comments for further details.

ankitrox commented:
Thanks @techanvil for in-depth review and providing the suggestions.

I have applied the recommended changes as per your suggestion and also removed the test file for `OverlayNotificationsRenderer` component as it was no longer necessary.

Also, updated the QAB over the issue, so assigning you for another review.

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationIntroductoryOverlayNotification.test.js:
```diff
@@ -50,22 +55,37 @@ describe( 'AudienceSegmentationIntroductoryOverlayNotification', () => {
 
 		expect(
-			getByText(
+			queryByText(
```
A trivial detail - when we _do_ expect an element to exist, we do use `getByText()` rather than `queryByText()`, as `getByText()` is slightly safer as it throws an error if an element with matching text isn't found.

```suggestion
			getByText(
```

Indeed, strictly speaking we don't even need to test for `toBeInTheDocument()` when using `getByText()` as the test will fail if the text isn't found. However we do have an informal convention of still checking for `toBeInTheDocument()`. I suppose it makes the test a bit more explicit.

techanvil commented on assets/js/modules/analytics-4/index.js:
```diff
@@ -172,12 +172,12 @@ export const registerWidgets = ( widgets ) => {
 				const audienceSegmentationSetupComplete =
```
Please rename this variable:

```suggestion
				const audienceSegmentationSetupCompletedBy =
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationIntroductoryOverlayNotification.js:
```diff
@@ -51,6 +55,17 @@ export default function AudienceSegmentationIntroductoryOverlayNotification() {
 	);
 
+	const shouldShowAudienceSegmentationIntroductoryOverlay = useSelect(
+		( select ) =>
+			select( CORE_MODULES ).isModuleActive( 'analytics-4' ) &&
+			( ! isViewOnly ||
+				select( CORE_USER ).canViewSharedModule( 'analytics-4' ) ) &&
+			select(
+				MODULES_ANALYTICS_4
+			).getAudienceSegmentationSetupCompletedBy() !==
+				select( CORE_USER ).getID()
```
This condition needs an additional test to ensure `audienceSegmentationSetupCompletedBy` is not `undefined` or `null`, to avoid showing the popup for the primary admin. I'd suggest using `Number.isInteger()` as follows:

```suggestion
		( select ) => {
			const isModuleActive =
				select( CORE_MODULES ).isModuleActive( 'analytics-4' );

			const canViewModule =
				! isViewOnly ||
				select( CORE_USER ).canViewSharedModule( 'analytics-4' );

			const audienceSegmentationSetupCompletedBy =
				select(
					MODULES_ANALYTICS_4
				).getAudienceSegmentationSetupCompletedBy();

			const userID = select( CORE_USER ).getID();

			return (
				isModuleActive &&
				canViewModule &&
				Number.isInteger( audienceSegmentationSetupCompletedBy ) &&
				audienceSegmentationSetupCompletedBy !== userID
			);
		}
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationIntroductoryOverlayNotification.js:
```diff
@@ -64,18 +79,20 @@ export default function AudienceSegmentationIntroductoryOverlayNotification() {
-			top: getNavigationalScrollTop( widgetClass, breakpoint ),
+			top: getNavigationalScrollTop( widgetAreaClass, breakpoint ),
 			behavior: 'smooth',
 		} );
```
In my testing, the notification is disappearing immediately without the page scrolling.

I've found that wrapping the call to `scrollTo()` in a `setTimeout()` with zero delay works:

```js
		setTimeout( () => {
			global.scrollTo( {
				top: getNavigationalScrollTop( widgetAreaClass, breakpoint ),
				behavior: 'smooth',
			} );
		}, 0 );
```

techanvil reviewed (changes requested):
Thanks @ankitrox. This needs a few more updates, please see my comments. Also, the QAB could use a small tweak as per my comment on the issue.
